pkg/userdata/translate/translate_v1alpha1.go

```diff
@@ -218,7 +218,7 @@ func translateV1Alpha1Init(nc *v1alpha1.NodeConfig, ud *userdata.UserData) error
 			APIVersion: "kubeadm.k8s.io/v1beta2",
 		},
 		ClusterName:          nc.Cluster.ClusterName,
-		KubernetesVersion:    constants.KubernetesVersion,
+		KubernetesVersion:    ud.KubernetesVersion,
```

Comment:
Since this is going to be user-supplied data, I would consider running it through a little bit of processing/validation.

At a minimum, I would `strings.TrimPrefix(ud.KubernetesVersion,"v")`.

Reply:
Yeah this is gonna be tricky. Upstream Kubernetes publishes images with the "v", and we can't always guarantee that if people use a custom registry (not currently possible) that the versions will be semantic with a "v" or no "v". For now we are pinned to the official GCR registry and they have the "v" in the tags. Any kind of validation you think we can/should do knowing this?

Reply:
This is something we deal with a lot in telephony.  (Do we dial a "1" or not?  Do we dial a "+" or not?  Do we dial a "9" or not?)

The best solution is generally to break it down into a root structure on the parsing side (strip the v and validate semver), and then format it on the use side (add a "v" if required by the system).

I'm not sure that translates well here since, as you note, the tags are actually free-form.

Never-the-less, if we call it "kubernetes version" as opposed to "container tag," I think there _should_ be a canonical form that is used internally.

Reply:
hmm... I think maybe requiring a valid semver would be nice? There is a library that has a `Parse` requiring strict semver.

Reply:
> I think maybe requiring a valid semver would be nice?
I'm hesitant to do that because it'd limit any end user built kubernetes. Should be the rare case and we don't currently support it ( given we don't flex on `constants.KubernetesImage` ).

Reply:
Yeah :/, its just if not semver then what is there to validate that we can depend on?

Reply:
Yeah, that's why this is a little messy.  The actual versions should be semver, but the container tags may not be.  It could be useful to offer both.  Among other things, we really will need to know the actual version for other things.

Perhaps we could add an option (later?) which would allow the user to override the computed container tag which still retaining a formal kubernetes version.

Reply:
^ sounds like a good start for now ... we have the rest of 0.3 to figure this out, and this at least unblocks the upgrade operator.

Reply:
is there harm in forcing users to use proper semantic versioning?

Reply:
I definitely see issues with that for companies which have their own add-ons and iterations.  Might work if they are forced into a particular structure, like:
`v1.16.1-mycompany-r13`

Reply:
fair point

Reply:
Regardless, none of this should hold up this PR.  I was just mentioning the very common quandary of `v1.16.1` vs `1.16.1` when dealing with user-supplied values.

Reply:
Which is to say that _I_ would generally strip the "v" on input and add it as necessary on output.

Reply:
I'm adding semver check for now, until we can figure out a better way. The good news is that due to talos CAPI changes, we can start running the validation code when generating configs.